java/org/apache/tomcat/util/http/FastHttpDateFormat.java

```diff
@@ -105,7 +105,7 @@ public final class FastHttpDateFormat {
     public static String getCurrentDate() {
         long now = System.currentTimeMillis();
         // Handle case where time moves backwards (e.g. system time corrected)
-        if (Math.abs(now - currentDateGenerated) > 1000) {
+        if (now - now%1000L != currentDateGenerated - currentDateGenerated%1000L) {
```

Comment:
There are faster ways of achieving the same results. This adds ~ 9% to getCurrentDate(). Look as dividing now by 1000 instead.

Reply:
Rough idea: add "current" variables currentDateGeneratedLower and currentDateGeneratedSecsUpper
Only set them when we generate a new currentDate, and then use:
currentDateGeneratedLower = currentDateGenerated - currentDateGenerated % 1000
currentDateGeneratedUpper = currentDateGeneratedLower + 1000
Finally change the new comparison expression to
if (now < currentDateGeneratedLower || now >= currentDateGeneratedUpper)
which involved only two comparisons in the hot path.
Untested, but I could try and provide a formal patch.

Reply:
to minimize costs , ignore ms part to avoid divide ops:
```java
    public static String getCurrentDate() {
        long now = System.currentTimeMillis();
        // Ignore millisecond part.
        now -= now % 1000L;
        // Handle case where time moves backwards (e.g. system time corrected)
        if (now != currentDateGenerated) {
            currentDate = FORMAT_RFC5322.format(new Date(now));
            currentDateGenerated = now;
        }
        return currentDate;
    }
```

Reply:
Modulus is also a slow operation.  If there is a chance of this slowing every request by 1ms as @markt-asf suggested then I'm a `-1` on this.  A `1ms` slowdown means 1,000 less requests-per-second in throughput.

Any unnecessary change should be benchmarked to evaluate the overhead that it adds.

Reply:
@isapir On my laptop pc, new implementation elapse ***3.32 nanoseconds / per call*** (single thread), comparing to original implementation's ***3.36 ns / per call***, even performance improved.

Please try it.

Reply:
@Chenjp If it's a few nanoseconds then that's fine by me.  How did you measure that?  Using JMH?

Reply:
TesterFastHttpDateFormatPerformance, change threadCount to 1.
result: Duration: 331058600 (***total callCount = 100000000***)
per call: 331058600 ns/ 100000000 (callCount) = ***3.31 ns***;